### DISCUSSION TITLE
How to implement soft delete

### DISCUSSION THREAD

**Discussion by donPuerto** [supabase/supabase#2799] [Questions] [answered]
I know there so many ways of doing this, but i want to know the best pattern in supabase?

thanks

**Comment by inian** [comment, marked as answer]
Hi @donPuerto, I personally prefer having a status column as a postgres enum type and setting it to "DELETED" for soft deletes.

**Comment by from-nibly** [comment]
I was coming to look for an answer for how to do RLS with supabase deletes. Generally speaking ~~best~~ a good practice with soft delete tables is to have several columns that support soft deletes.

### minimum

deleted_at: timestamp (nullable)

### bonus

deleted_by: user_id (nullable)

### index performance (but also not normalized)

deleted: boolean

* can be nullable in order to make trinary state
null = never been deleted
true = deleted
false = undeleted

* alternatively you can leave deleted_at and deleted_by populated when going from true to false to indicate that it was previously deleted.

Going to look at some RLS for this to prevent deletes from all users, and to specifically limit who can "delete" or "undelete" objects, as well as who can even see deleted objects

There is also stuff like this where you move the row to another table, though that means you have to perform all migrations on two tables every time you modify a given table.

https://stackoverflow.com/questions/41752660/how-to-properly-implement-soft-delete-in-postgresql

**Comment by MMTE** [comment]
maybe worths considering:
https://brandur.org/fragments/deleted-record-insert

**Comment by TheDistanceHQ** [comment]
We've just created a SQL implementation of the official Go function

https://medium.com/@TheDistance/supabase-soft-delete-via-sql-6e4f278b4236
